### ISSUE TITLE
Square Roots in Fields of Prime Power Order

### ISSUE THREAD

**Issue by TDQuering** [mhostetter/galois#573]
Greetings!

It appears the functionality with `np.sqrt` to compute square roots of elements is not properly implemented for extension fields of prime fields. I ran into this issue while trying to compute the square root of 6 in GF(11^2), which incorrectly returns a value of 1. I did a little more testing, and I found that the package computes the square root of *all* nonzero elements of GF(11^2) as 1. After some further investigation, it seems to be giving incorrect square root results consistently in fields GF(p^k) with odd characteristic p and even extension degree k.

It makes sense that there are no issues in characteristic 2, as the square root method for binary fields uses a specific implementation that appears correct: 
https://github.com/mhostetter/galois/blob/c945356533b369adc4860e7979c34b04c862ba1d/src/galois/_domains/_calculate.py#L758-L768

I believe the reason we have problems in the odd characteristic case is because the algorithms referenced here are designed for computing square roots modulo p, i.e. in the prime field, but not necessarily in extension fields: 
https://github.com/mhostetter/galois/blob/c945356533b369adc4860e7979c34b04c862ba1d/src/galois/_domains/_calculate.py#L770-L779

It's interesting that these algorithms appear (under admittedly limited testing) to work in extension fields of odd degree. Ideally, it's probably worth verifying that this is a genuine mathematical property and not just a fluke due to the small amount of testing I've done to research this issue. But it appears a different technique is definitely required to handle extension fields of even extension degree. I've found [a paper ](https://eprint.iacr.org/2012/685) on the problem of square roots in finite fields of even extension degree which looks promising, but I haven't had the chance to fully investigate it myself.

Also, I just wanted to say that I really appreciate those who have worked on this package - it's been very helpful to me. I hope this can be used to improve it even more.

**Comment by mhostetter** [comment]
I can reproduce this bug.

```python
import galois

GF = galois.GF(11**2)
x = GF.squares
print(x)
y = np.sqrt(x)
print(y)
print(y**2)
```

```
[  0   1   2   3   4   5   6   7   8   9  10  13  14  15  16  20  26  28
  29  30  32  34  37  38  39  42  45  47  49  52  53  56  58  59  64  65
  67  68  73  74  76  79  80  83  85  87  90  93  94  95  98 100 102 103
 104 106 112 116 117 118 119]
[0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
[0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
```

**Comment by mhostetter** [comment]
Thank you for identifying this bug and doing as much debug as you already have.

From [the textbook](https://cacr.uwaterloo.ca/hac/about/chap3.pdf) I referenced, it says

![image](https://github.com/user-attachments/assets/1920f06f-132d-43a9-88a2-f194dea3bc0b)

Perhaps I didn't "straightforwardly" extend the algorithms to extension fields? Maybe there's a bug in my Python code?

**Comment by mhostetter** [comment]
This was the first thing I tried, but apparently I didn't apply the idea correctly. I just re-tried, with your patch, and it does seem to work. Thanks for debugging this!

I can submit a PR soon, or you can if interested.

**Comment by TDQuering** [comment]
Done! See the PR here: https://github.com/mhostetter/galois/pull/574

**Comment by mhostetter** [comment]
Released in 0.4.3.

**Comment by TDQuering** [comment]
Just updated to the new release and things are working as intended! Thanks again for all the work you put into this package. :-)
